### DISCUSSION TITLE
ISMAGs (Sandia Lab) - Mentored Project [GSoC 23]

### DISCUSSION THREAD

**Discussion by jnash10** [networkx/networkx#6532] [Algorithms]
Hi,

As mentioned in the mentored projects page (https://networkx.org/documentation/latest/developer/projects.html#incorporate-a-python-library-for-ismags-isomorphism-calculations) , comparing the current implementation of the ISMAGs algorithms with the one provided to NetworkX by Sandia Labs is one of the mentioned issues.

This issue is also one of the GSoC project tracks for NetworkX.

If possible, could I get access to the Sandia Labs code or other resources to get started!
I'd love to take this up as my GSoC'23 application.

Dear @dschult, in case the Sandia Labs code is not to be shared till GSoC starts, what would be a good direction to start in? I'm currently going through the current ISMAGs code https://github.com/networkx/networkx/blob/126a06c6315d9bcbc3b939d87f1764f77ed00cd8/networkx/algorithms/isomorphism/ismags.py and the isomorphism API in NetworkX in general and will post my updates and understanding of the algorithm here if that's okay. 

Would really appreciate discussion and feedback! 
Thanks a lot.

**Comment by dschult** [comment]
That's a good way to start -- learning the current networkx implementation of ISMAGS and the networkx isomorphism API more generally.  The Sandia Labs version is much more complete.

The Sandia Labs ISMAGS code is on github:  [https://github.com/sandialabs/ISMAGS-in-Python](https://github.com/sandialabs/ISMAGS-in-Python)

**Comment by jnash10** [comment]
Hi again,

I read the ISMAGs paper and have been scouring through NetworkX's and Sandia Lab's ISMAGs working on understanding it.

I totally get why it has been coined as a 'Full time 10 weeks' project.

Is there any specific order in which to pick up and learn things that you'd recommend so that I can get a better handle on things?

My contributor guideline workflow is set up so I can start with any small issue too if that'd help in any way!

I saw that you @pckroon worked on the original implementation. What was the process you guys followed?

**Comment by pckroon** [comment]
Oooh, interesting project! (and hard!)

I reimplemented ISMAGS based on the paper, without using the java implementation as reference to avoid any licensing issues. I needed it for https://github.com/marrink-lab/vermouth-martinize, and after that realized it might also be a nice addition to nx, and opened a PR.
There are definitely things in my implementation that could be improved. For example, my version creates the node compatibility dictionary between `graph` and `subgraph` a little bit strangely by first sorting the nodes in graph and subgraph in partitions, and assessing compatibility between the partitions (https://github.com/networkx/networkx/blob/main/networkx/algorithms/isomorphism/ismags.py#L362). For `subgraph` this is fine (you need it anyway for the symmetry), but between the `graph` and `subgraph` it's a bit silly, since it results in doing `O(m*m)` comparisons rather than `O(n*m)` (and `m>=n`) with `n=len(subgraph)` and `m=len(graph)`. 
Another point where I *know* I deviate from the reference implementation is https://github.com/networkx/networkx/blob/main/networkx/algorithms/isomorphism/ismags.py#L874 to 879. I use sets, the reference uses sorted lists to more quickly filter out all nodes smaller/larger. A sorted-set would be the ideal compromise here.
Another point where I think are differences is in https://github.com/networkx/networkx/blob/main/networkx/algorithms/isomorphism/ismags.py#L784 --- no idea how the original solved this.

My 10000-mile-height-I-last-looked-at-this-code-at-least-4-years-ago take is that my `_map_nodes` method and generally the interface are pretty decent, but that a lot can be gained in particular in the symmetry analysis: start by reviewing/replacing `_process_ordered_pair_partitions`.

If you have specific questions about my implementation I would be more than happy to try and help, but as mentioned it's been a while since I looked at this.

**Reply by jnash10** [reply]
Thank you so much for taking the time.

Yeah, I'll probably be working on this as my summer project. 
Having it selected for GSoC would be added motivation :)

**Reply by jnash10** [reply]
Hi @pckroon, please correct me if I'm wrong,

Do you mean partition and comparison between ```graph``` and ```graph``` which is ```O(m*m)``` for the first issue/improvement?
Since comparison between ```subgraph``` and ```graph``` is ```O(m*n)``` and is anyway required.

Do excuse me if I made an error in interpretation somewhere.

**Reply by pckroon** [reply]
Right.
At the time I thought it would be clever to group all nodes of `graph` and `subgraph` into sets that contain equivalent nodes (using `make_partitions`). Then, to assess subgraph-graph node compatibility I only had to look at compatibility between representatives of those partitions. If I remember correctly you do need to make those partitions for `subgraph` for the symmetry detection. 
Best case making those partitions is `O(1)` (everything is compatible), worst case it's `O(len(graph)*len(graph))` which is terrible.

**Comment by jnash10** [comment]
Dear, @dschult 

Since the deadline for proposal application is approaching, shall I start with a proposal?
I'd like to base my plan on the pointers given by @pckroon. Mainly focussing on how symmetries are detected and speeding up the subroutines.

Please let me know your thoughts regarding the same if possible.

**Reply by MridulS** [reply]
Please do! The deadline is approaching quickly :)

**Comment by jnash10** [comment]
Dear @dschult and @MridulS,

Could you please take a few minutes to review my proposal for GSoC'23 ?
[link to the proposal](https://github.com/jnash10/NetworkX-GSoC-23/blob/main/gsoc_proposal_agamdeep.pdf)

I'd be grateful to hear your suggestions and feedback.
Thank you for taking the time!

**Reply by dschult** [reply]
This looks reasonable in terms of what is there and how it is written. You should check the spelling/typos.

**Reply by jnash10** [reply]
> This looks reasonable in terms of what is there and how it is written. You should check the spelling/typos.

Yup, Will sort the typos.

Thanks for taking the time!

**Comment by jnash10** [comment]
Updated proposal: [PDF](https://github.com/jnash10/NetworkX-GSoC-23/blob/0780ae04eb74310fbd326c5b68327e7d777adafd/gsoc_2023_networkx_ISMAGS_agamdeep.pdf)

Changes:

- Added testing section
- moved profile to after project details
- reduced margins
- increased line spacing
